Libplanet/Blockchain/BlockChain.cs

```diff
@@ -208,11 +228,49 @@ IEnumerator IEnumerable.GetEnumerator()
 
             foreach (var hashValue in hashValues)
             {
+                AddressStateMap blockStates = Store.GetBlockStates(hashValue);
+                if (blockStates is null)
+                {
+                    if (completeStates)
+                    {
+                        // Calculates and fills the incomplete states
+                        // on the fly.
+                        foreach (Block<T> b in this)
+                        {
+                            if (!(Store.GetBlockStates(b.Hash) is null))
+                            {
+                                continue;
+                            }
+
+                            ActionEvaluation<T>[] evaluations =
+                                b.Evaluate(
+                                    DateTimeOffset.UtcNow,
+                                    a => GetStates(
+                                        new[] { a },
+                                        b.PreviousHash
+                                    ).GetValueOrDefault(a)
+                                ).ToArray();
+                            SetStates(b, evaluations);
```

Comment:
I think `StateReference` and `TxNonce` should be cleared before call `SetStates`. Because they were made on the assumption that it is stored in order.

Reply:
I wrote a test code for the case you mentioned, and it indeed failed.  I fixed `GetStates()` not to build indices (`TxNonce` and `StateReference`): 8708fd1de8127a2753f77175ca02b5f83662dbb9.